tensorboard/plugins/graph/graph_util.py

```diff
@@ -0,0 +1,214 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for graph plugin."""
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.tensorflow_stub import dtypes
+
+
+def _walk_layers(keras_layer):
+  """Walks the nested keras layer configuration in preorder.
+  Args:
+    keras_layer: Keras configuration from model.to_json.
+
+  Yields:
+    (name_scope, layer_config)
+  """
+  yield ('', keras_layer)
+  if keras_layer.get('config').get('layers'):
+    name_scope = keras_layer.get('config').get('name')
+    for layer in keras_layer.get('config').get('layers'):
+      for (sub_name_scope, sublayer) in _walk_layers(layer):
+        sub_name_scope = '%s/%s' % (
+            name_scope, sub_name_scope) if sub_name_scope else name_scope
+        yield (sub_name_scope, sublayer)
+
+
+def _scoped_name(name_scope, node_name):
+  """Returns scoped name for a node.
+
+  Args:
+    name_scope: scope names similar to that of tf.name_scope.
+    node_name: current node name.
+
+  Returns
+    A scoped name.
+  """
+  if name_scope:
+    return '%s/%s' % (name_scope, node_name)
+  return node_name
+
+
+def _is_model(layer):
+  """Returns true if layer is a model."""
+  return layer.get('config').get('layers') is not None
+
+
+def _norm_to_list_of_layers(maybe_layers):
+  """Normalizes to a list of layers.
+
+  A Functional model has fields 'inbound_nodes' and 'output_layers'
+  which looks like below.
+  - ['in_layer_name', 0, 0, {}] # [name, size, index, unknown]
+  - [['in_layer_is_model', 1, 0, {}], ['in_layer_is_model', 1, 1, {}]]
+
+  Normalize to always return a list of item.
+  """
+  return (maybe_layers if isinstance(maybe_layers[0], (list,))
+          else [maybe_layers])
+
+
+def _update_dicts(name_scope,
+                  model_layer,
+                  input_to_in_layer,
+                  model_name_to_output,
+                  prev_node_name):
+  """Updates input_to_in_layer, model_name_to_output, and prev_node_name
```

Comment:
How does it update them?

Reply:
How would you like me to improve name of the func and the comment? It side-effects the dicts and returns them at the end/